### ISSUE TITLE
Use Average True Range to determine the size of a position to take out on a trade

### ISSUE THREAD

**Issue by bvkin** [bvkin/trade_bot#36]
[Average True Range ](https://www.investopedia.com/terms/a/atr.asp#:~:text=An%20average%20true%20range%20value,of%20%241.18%20per%20trading%20day.) can be used to determine market volatility for a stock. This may be used to determine the size of a position to take out on a purchase.

**Comment by bvkin** [comment]
FYI - Average True Range can be calculated by [talib](https://github.com/TA-Lib/ta-lib-python/blob/master/docs/func_groups/volatility_indicators.md#atr---average-true-range)

**Comment by bbell03** [comment]
Average True Range


Purpose: Incorporate ATR into portfolio allocation on buy

Ranges based on risk tolerance so…

Use ATR to indicate
	[low risk]
		low risk procedure
	[medium risk]
		medium risk procedure
	[high risk]
		high risk procedure

on notional order amount apply a given modifier based on ATR (atr or incorporated metric)

**Comment by bvkin** [comment]
Let's create a new file `utils/trading_calculators.py` and in it add the following functions:
`calculate_purchase_amnt(buying_power, buying_percentage)` - This will return the dollar amount to purchase of the given stock. `buying_power` should be the total buying power available in the alpaca account,`buying_percentage` should be the percent of the total buying power we're willing to stake in the trade. Right now this is hard coded at `0.05`, but let's instead make this configurable as an environment variable.

**Comment by bbell03** [comment]
Ok I think maybe we spin this into either a util on another class or expand the class to include different methods for allocation.

I'll push something today to start.

**Comment by bbell03** [comment]
Proposed class outline:

`class TradingCalculator:
    def __init__(self, trade_manager: AlpacaTradeManager, Strategy: Strategy, sns_client: SNSClient,  sns_topic_arn: str = None):
        self.trade_manager = trade_manager
        self.strategy = Strategy
        self.client: sns_client
        self.topic: sns_topic_arn
        self.multiplier = os.environ["VARIABLE_NAME"] = "value"
        self.atr

    def calculate_purchase_amnt(self, buying_power: float, buying_percentage: float) -> int:
    def calculate_sl_tp(self, purchase_amnt: float, sltp_multiplier: float, df:pd.DataFrame) -> Tuple:
         return [self.calculate_stop_loss(purchase_amnt, sltp_multiplier), self.calculate_take_profit(purchase_amnt, sltp_multiplier)]
    def calculate_stop_loss(self, purchase_amnt: float, sl_multiplier: float, df:pd.DataFrame)-> float:
        return purchase_amnt - (self.atr[-1] * self.multiplier) 
    def calculate_take_profit(self, purchase_amnt: float, tp_multiplier: float, df:pd.DataFrame) -> float:
        return purchase_amnt + (self.atr[-1] * self.multiplier)

       def get_price_data(self, ticker: str, period_start: str, period_end: str, timeframe: str = "day", adjustment: str = 'raw') -> pd.DataFrame:
        return self.trade_manager.get_price_data(ticker, period_start, period_end, timeframe, adjustment)

       def get_atr(self, df: pd.DataFrame) -> float:

`

**Comment by bvkin** [comment]
Hey @bbell03 this is going to expand the scope of this issue quite a bit. Could we stick to the original outline to get this issue closed?

**Comment by bbell03** [comment]
Yeah that sounds fine for now I guess but was thinking in terms of when and where we might need to calculate amounts for trading generally across the app.

For example we may want similar functionality for the following use cases:

**Comparing across 2 or more trades given current state of the portfolio**

_- What would buying AAPL look like right now vs buying MSFT?_

**Simulating transactions on historical data and simulated portfolio for backtesting?**

_- What if we bought/sold this ticker at this time with a given allocation under given portfolio constraints?_

Seems like both of these use cases would fall under the same functionality or have significant overlap. I'm wondering what you think about performing trade calculations elsewhere in our app.

It might be out of scope of the current issue but we might want to take these kinds of considerations into account so the app is easier to dev on scale and integrate moving forward.

For the current issue I think it's fine if I just implement the changes in your spec but I wonder how we could make our lives easier for the future and find common ground across modules.

**Comment by bvkin** [comment]
Us [this](https://github.com/bvkin/trade_bot/blob/main/trade_bot/main.py#L41-L42) as reference.

**Comment by bvkin** [comment]
> Yeah that sounds fine for now but was thinking in terms of when and where we might need to calculate amounts and stats for trading generally across the app.
> 
> For example we may want similar functionality for the following use cases:
> 
> **Comparing across 2 or more trades given current state of the portfolio**
> 
> _- What would buying AAPL look like right now vs buying MSFT?_
> 
> **Simulating transactions on historical data and simulated portfolio for backtesting**
> 
> _- What if we bought/sold this ticker at this time with a given allocation and portfolio constraints?_
> 
> To me it seems like both of these use cases would fall under the same TradingCalculator functionality or have significant overlap. I'm wondering what you think about performing trade calculations elsewhere in the app.
> 
> It might be out of scope of the current issue but it could be helpful to take these kinds of considerations into account so the app is easier to dev on scale and integrate moving forward.
> 
> For the current issue I think it's fine if I just implement the changes in your spec but I wonder how we could make our lives easier for the future and find common ground across modules.

trading_calculator should just be a utility library. These other things should be or are already implemented elsewhere.
